### PR TITLE
Feature/daemonset support

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [ ] Do comments answer the question of why design decisions were made?

----

## Description of change

Add daemon set support;

## QA steps

deploy CaaS application with `deployment.type` == `daemon` in `metadata.yaml`;

```yaml
# metadata.yaml
deployment:
  type: daemon
```

```console
$ kubectl get ds -n t1
NAME          DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
mariadb-k8s   1         1         0       1            0           <none>          4m6s

$ juju remove-application mariadb-k8s
removing application mariadb-k8s
- will detach storage database/0

$ kubectl get ds/mariadb-k8s -n t1
Error from server (NotFound): daemonsets.apps "mariadb-k8s" not found

$ juju deploy cs:~kelvin.liu/mariadb-k8s-x --constraints tags=kubernetes.io/hostname=ubuntu

$ kubectl get ds/mariadb-k8s -n t1 -o jsonpath="{..affinity}"
map[nodeAffinity:map[requiredDuringSchedulingIgnoredDuringExecution:map[nodeSelectorTerms:[map[matchExpressions:[map[key:kubernetes.io/hostname operator:In values:[ubuntu]]]]]]]]

$ juju remove-application mariadb-k8s
removing application mariadb-k8s
- will detach storage database/0

$ kubectl get ds/mariadb-k8s -n t1
Error from server (NotFound): daemonsets.apps "mariadb-k8s" not found

```

## Documentation changes

Yes

## Bug reference

None

### CODE REVIEW

**Comment by ycliuhw** [comment]
[k8s Daemonset support](https://trello.com/c/KqfvQVua/1042-k8s-daemonset-support)

File: caas/kubernetes/provider/k8s.go [hunk 1]
```diff
@@ -1048,39 +1056,24 @@ func (k *kubernetesClient) EnsureService(
...
-	existingStatefulSet, err := statefulsets.Get(deploymentName, v1.GetOptions{IncludeUninitialized: true})
-	if err != nil && !k8serrors.IsNotFound(err) {
-		return errors.Trace(err)
-	}
-	if !useStatefulSet {
-		useStatefulSet = err == nil
-		if useStatefulSet {
-			logger.Debugf("no updated filesystems but already using stateful set for %v", appName)
+	if params.Deployment.DeploymentType == "" {
+		// TODO(caas): we should really change `params.Deployment` to be required.
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: The current behaviour is to default to stateless if there's no storage definitions and stateful otherwise. I'm not sure that's a bad thing. Most people will have no need to worry about setting deployment type.

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -1118,20 +1111,30 @@ func (k *kubernetesClient) EnsureService(
...
 		cleanups = append(cleanups, func() { k.deleteDeployment(appName) })
+	case caas.DeploymentDaemon:
+		cleanUpDaemonSet, err := k.configureDaemonSet(appName, deploymentName, annotations.Copy(), workloadSpec, params.PodSpec.Containers)
+		if err != nil {
+			return errors.Annotate(err, "creating or updating DaemonSet")
+		}
+		cleanups = append(cleanups, cleanUpDaemonSet)
+	default:
+		// This should never happend because we have validated both in this method and in `charm.v6`.
+		return errors.NotSupportedf("deployment type %q not supported", params.Deployment.DeploymentType)
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: NotSupportedf() already appends "not supported"

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -1601,14 +1663,38 @@ func (k *kubernetesClient) ensureStatefulSet(spec *apps.StatefulSet, existingPod
...
-	_, err = api.Update(existing)
+	_, err = k.updateStatefulSet(existing)
 	return errors.Trace(err)
 }
 
-// createStatefulSet deletes a statefulset resource.
-func (k *kubernetesClient) createStatefulSet(spec *apps.StatefulSet) error {
-	_, err := k.client().AppsV1().StatefulSets(k.namespace).Create(spec)
-	return errors.Trace(err)
+func (k *kubernetesClient) createStatefulSet(spec *apps.StatefulSet) (*apps.StatefulSet, error) {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: as a driveby can we separate these out to a new statefulset.go file like was done for daemonset
(plus the delete, list etc)

File: caas/kubernetes/provider/k8s_test.go [hunk 4]
```diff
@@ -3277,6 +3289,186 @@ func (s *K8sBrokerSuite) TestEnsureServiceForDeploymentWithDevices(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 }
 
+func (s *K8sBrokerSuite) TestEnsureServiceForDaemonSetWithDevicesCreate(c *gc.C) {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: Can we include a test with constraint tags so that we test that node affinity is set etc

**Comment by ycliuhw** [comment]
$$merge$$
